### PR TITLE
Extended error message in case gcc-xml existed unsuccessfull

### PR DESCRIPTION
**Pull request by goldhoorn**
The old-error message was often-non-speaking because the error got hidden.
I first thought this should be only displayed when calling --debug or --verbose but 
sometimes the headers has typos and the error sometimes only display the most top level one.
Even gcc-xml itself has problems this would make the problem more-understandable.

Thougts @doudou @marvin2k ?

### CODE REVIEW

**Comment by marvin2k** [comment]
Hm. Isn't the gccxml-output more interesting? Adapting the outputted "command" to contain the path to the copied debug-file would allow copy-pasting it into a shell. Ant note that there is a race when two autoprojects fail at the same time.

**Comment by doudou** [comment]
The reason why nothing was shown is that gccxml's error output is not redirected, i.e. is displayed to the user. We could simply say that ...

**Comment by goldhoorn** [comment]
Say what?

**Comment by doudou** [comment]
Not sure what I wanted to say, honestly. My feeling is that the gccxml/clang error output should be enough.

**Comment by goldhoorn** [comment]
Ok, let's try this one.
Mix between your's and mine.

Btw. why we not using $DEBUG in more project, why you specify your own debug flags?

**Comment by doudou** [comment]
> Btw. why we not using $DEBUG in more project, why you specify your own debug flags?

Because, otherwise, when setting $DEBUG you would get the debug output of *all* the libraries you use. Let's try to debug a typelib issue while running syskit ... oh wait, can't.

There's a per-namespace logger (in typelib, it is on Typelib), and a Logger-compatible interface (i.e. 

~~~
Typelib.debug do
   ... do stuff ...
   "message"
end
~~~

**Comment by goldhoorn** [comment]
updated

File: bindings/ruby/lib/typelib/gccxml.rb [hunk 1]
```diff
@@ -830,13 +830,24 @@ def self.preprocess(files, kind, options)
...
+                        gccxml_io.read
+                    end
+                    if !$?.success?
+                        Typelib.debug do
+                            FileUtils.copy_file(io.path, "/tmp/gcc-debug")
+                            "Command was \"#{command.join(" ")}\".
+    Debug file created at \"/tmp/gccxml-debug\"."
+                        end
+                        raise ArgumentError, "gccxml failed to preprocess #{files.join(" ")}:
+\"
```
Comment (doudou) [inline bindings/ruby/lib/typelib/gccxml.rb]: Why the quotes ? It's really ugly

File: bindings/ruby/lib/typelib/gccxml.rb [hunk 2]
```diff
@@ -830,13 +830,24 @@ def self.preprocess(files, kind, options)
...
-                    raise ArgumentError, "failed to preprocess #{files.join(" ")}"
+                Tempfile.open('gccxml_err') do |err|
+                    command = ["gccxml", "--preprocess", *includes, *defines, *gccxml_default_options, io.path]
+                    result = IO.popen(command, :err => err.path) do |gccxml_io|
+                        gccxml_io.read
+                    end
+                    if !$?.success?
+                        Typelib.debug do
+                            FileUtils.copy_file(io.path, "/tmp/gcc-debug")
+                            "Command was \"#{command.join(" ")}\".
```
Comment (doudou) [inline bindings/ruby/lib/typelib/gccxml.rb]: This breaaks logger formatting. If you want multiple lines, do

~~~
Typelib.debug do
   Typelib.debug "line0"
   Typelib.debug "line1"
   break
end
~~~

File: bindings/ruby/lib/typelib/gccxml.rb [hunk 3]
```diff
@@ -830,13 +830,24 @@ def self.preprocess(files, kind, options)
...
                 end
                 io.flush
-                result = IO.popen(["gccxml", "--preprocess", *includes, *defines, *gccxml_default_options, io.path]) do |gccxml_io|
-                    gccxml_io.read
-                end
-                if !$?.success?
-                    raise ArgumentError, "failed to preprocess #{files.join(" ")}"
+                Tempfile.open('gccxml_err') do |err|
+                    command = ["gccxml", "--preprocess", *includes, *defines, *gccxml_default_options, io.path]
+                    result = IO.popen(command, :err => err.path) do |gccxml_io|
```
Comment (doudou) [inline bindings/ruby/lib/typelib/gccxml.rb]: Use keyword argument syntax

~~~
err: err.path
~~~

**Comment by goldhoorn** [comment]
updated

File: bindings/ruby/lib/typelib/gccxml.rb [hunk 4]
```diff
@@ -830,13 +830,24 @@ def self.preprocess(files, kind, options)
...
+                        gccxml_io.read
+                    end
+                    if !$?.success?
+                        Typelib.debug do
+                            FileUtils.copy_file(io.path, "/tmp/gcc-debug")
+                            Typelib.debug "Command was '#{command.join(' ')}."
+                            Typelib.debug "Debug file created at '/tmp/gccxml-debug'."
+                        end
+                        raise ArgumentError, "gccxml failed to preprocess #{files.join(' ')}:
+'
```
Comment (doudou) [inline bindings/ruby/lib/typelib/gccxml.rb]: Still no idea why you absolutely want to put quotes here ...

Comment (goldhoorn) [inline bindings/ruby/lib/typelib/gccxml.rb]: To make clear that his is a sub-output

Comment (doudou) [inline bindings/ruby/lib/typelib/gccxml.rb]: Then add a line saying so explicitely above it:
~~~
gccxml output:
#{gccxml_output}
~~~

The quotes are super ugly, and don't make it clear at all (at least, not to me)

**Comment by goldhoorn** [comment]
updated

**Pull request merged**